aptos-move/move-examples/token_objects/sources/property_map.move

```diff
@@ -0,0 +1,627 @@
+            U32
+        } else if (type == string::utf8(b"u64")) {
+            U64
+        } else if (type == string::utf8(b"u128")) {
+            U128
+        } else if (type == string::utf8(b"u256")) {
+            U256
+        } else if (type == string::utf8(b"address")) {
+            ADDRESS
+        } else if (type == string::utf8(b"vector<u8>")) {
+            BYTE_VECTOR
+        } else if (type == string::utf8(b"0x1::string::String")) {
+            STRING
+        } else {
+            abort(error::invalid_argument(ETYPE_INVALID))
+        }
+    }
+
+    inline fun validate_type(type: u8, value: vector<u8>) {
+        if (type == BOOL) {
+            from_bcs::to_bool(value);
+        } else if (type == U8) {
+            from_bcs::to_u8(value);
+        } else if (type == U16) {
+            from_bcs::to_u16(value);
+        } else if (type == U32) {
+            from_bcs::to_u32(value);
+        } else if (type == U64) {
+            from_bcs::to_u64(value);
+        } else if (type == U128) {
+            from_bcs::to_u128(value);
+        } else if (type == U256) {
+            from_bcs::to_u256(value);
+        } else if (type == ADDRESS) {
+            from_bcs::to_address(value);
+        } else if (type == BYTE_VECTOR) {
+            // nothing to validate...
+        } else if (type == STRING) {
+            from_bcs::to_string(value);
+        } else {
+            abort(error::invalid_argument(ETYPE_INVALID))
+        };
+    }
+
+    public fun generate_mutator_ref(ref: &ConstructorRef): MutatorRef {
+        MutatorRef { self: object::address_from_constructor_ref(ref) }
+    }
+
+    // Accessors
+
+    public fun contains_key<T: key>(object: &Object<T>, key: &String): bool acquires PropertyMap {
+        assert!(
+            exists<PropertyMap>(object::object_address(object)),
+            error::not_found(EPROPERTY_MAP_DOES_NOT_EXIST),
+        );
+        let property_map = borrow_global<PropertyMap>(object::object_address(object));
+        simple_map::contains_key(&property_map.inner, key)
+    }
+
+    public fun length<T: key>(object: &Object<T>): u64 acquires PropertyMap {
+        assert!(
+            exists<PropertyMap>(object::object_address(object)),
+            error::not_found(EPROPERTY_MAP_DOES_NOT_EXIST),
+        );
+        let property_map = borrow_global<PropertyMap>(object::object_address(object));
+        simple_map::length(&property_map.inner)
+    }
+
+    public fun read<T: key>(object: &Object<T>, key: &String): (String, vector<u8>) acquires PropertyMap {
+        assert!(
+            exists<PropertyMap>(object::object_address(object)),
+            error::not_found(EPROPERTY_MAP_DOES_NOT_EXIST),
+        );
+        let property_map = borrow_global<PropertyMap>(object::object_address(object));
+        let property_value = simple_map::borrow(&property_map.inner, key);
+        let new_type = to_external_type(property_value.type);
+        (new_type, property_value.value)
+    }
```

Comment:
You can make this an inline

Reply:
the problem of making everything inline is upgradeability. I'm not actually a fan of public inline because it breaks the concept of upgradeability. if A uses B::inline, an upgrade to B::inline to B::inline2 still results in A calling B::inline.